Reorganize CONTRIBUTING.md (#9281)
matrix-org/synapse: CONTRIBUTING.md

Welcome to Synapse

This document aims to get you started with contributing to this repo! 

> - [1. Who can contribute to Synapse?](#1-who-can-contribute-to-synapse-)
> - [2. What do I need?](#2-what-do-i-need-)

Reviewer: These links are broken:

```suggestion
- [1. Who can contribute to Synapse?](#1-who-can-contribute-to-synapse)
- [2. What do I need?](#2-what-do-i-need)
```

- [3. Get the source.](#3-get-the-source)
- [4. Install the dependencies](#4-install-the-dependencies)

>   * [Under Unix (macOS, Linux, BSD, ...)](#under-unix--macos--linux--bsd---)

Reviewer: ```suggestion
  * [Under Unix (macOS, Linux, BSD, ...)](#under-unix-macos-linux-bsd-)
```

  * [Under Windows](#under-windows)
- [5. Get in touch.](#5-get-in-touch)
- [6. Pick an issue.](#6-pick-an-issue)

> - [7. Turn coffee and documentation into code and documentation!](#7-turn-coffee-and-documentation-into-code-and-documentation-)
> - [8. Test, test, test!](#8-test--test--test-)

Reviewer: ```suggestion
- [7. Turn coffee and documentation into code and documentation!](#7-turn-coffee-and-documentation-into-code-and-documentation)
- [8. Test, test, test!](#8-test-test-test)
```

  * [Run the linters.](#run-the-linters)
  * [Run the unit tests.](#run-the-unit-tests)
  * [Run the integration tests.](#run-the-integration-tests)
- [9. Submit your patch.](#9-submit-your-patch)
  * [Changelog](#changelog)

>     + [How do I know what to call the changelog file before I create the PR?](#how-do-i-know-what-to-call-the-changelog-file-before-i-create-the-pr-)

Reviewer: ```suggestion
    + [How do I know what to call the changelog file before I create the PR?](#how-do-i-know-what-to-call-the-changelog-file-before-i-create-the-pr)
```

license - in our case, this is almost always Apache Software License v2 (see
[LICENSE](LICENSE)).

# 2. What do I need?

The code of Synapse is written in Python 3. To do pretty much anything, you'll need [a recent version of Python 3](https://wiki.python.org/moin/BeginnersGuide/Download).

The source code of Synapse is hosted on Github. You will also need [a recent version of git](https://github.com/git-guides/install-git).

For some tests, you will need [a recent version of Docker](https://docs.docker.com/get-docker/).

## Under Unix (macOS, Linux, BSD, ...)

Once you have installed Python 3, please open a Terminal and run:

```sh
mkdir -p ~/synapse
python3 -m venv ./env
source ./env/bin/activate

> pip install -e ".[all,lint,mypy,test,black,tox]"

Reviewer: ```suggestion
pip install -e ".[all,lint,mypy,test]"
```

black and tox aren't valid identifiers here. black is installed a part of the `lint` target, maybe you meant to install tox separately?

```

This will install the developer dependencies for the project.

## Under Windows

TBD

# 3. Get the source.

The preferred and easiest way to contribute changes is to fork the relevant
project on github, and then [create a pull request](
https://help.github.com/articles/using-pull-requests/) to ask us to pull your
changes into our repo.

Please base your changes on the `develop` branch.

```sh
git clone git@github.com:matrix-org/synapse.git

> ```

Reviewer: I think this actually gives the wrong advice since it won't pull from their fork?

# 4. Get in touch.

Join our developer community on Matrix: #synapse-dev:matrix.org !

# 5. Pick an issue.

> Fix your favorite problem of find an issue on https://github.com/matrix-org/synapse/issues/ .

Reviewer: ```suggestion
Fix your favorite problem or [find an issue](https://github.com/matrix-org/synapse/issues/). You might want to take a look at one of the [Good First Issues](https://github.com/matrix-org/synapse/issues?q=is%3Aopen+is%3Aissue+label%3A%22Good+First+Issue%22).
```

# 7. Test, test, test!
<a name="test-test-test"></a>

While you're developing and before submitting a patch, you'll
want to test your code.

## Run the linters.

The linters look at your code and do two things:

- ensure that your code follows the coding style adopted by the project;
- catch a number of errors in your code.

They're very fast, don't hesitate!

```sh
source ./env/bin/activate
./scripts-dev/lint.sh
```

**Note that the script does not just test/check, but also reformats code, so you
may wish to ensure any new code is committed first**.
Note that the linters *will modify your files* to fix styling errors.
Make sure that you have saved all your files.

By default, this script checks all files and can take some time; if you alter
only certain files, you might wish to specify paths as arguments to reduce the
run-time:

```
If you wish to restrict the linters to only run on some files, you can instead run:

```sh
source ./env/bin/activate
./scripts-dev/lint.sh path/to/file1.py path/to/file2.py path/to/folder
```

You can also provide the `-d` option, which will lint the files that have been
changed since the last git commit. This will often be significantly faster than

> linting the whole codebase.

Reviewer: This seems like a useful comment to keep.

> If you wish to only run *some* unit tests, you may specify
> another module instead of `tests`:
> 
> ```sh
> source ./env/bin/activate
> python -m twisted.trial tests.rest.admin.test_room
> ```

Reviewer: Might be worth mention that you can specify one or more modules / class / method.

```suggestion
If you wish to only run *some* unit tests, you may specify another module
instead of `tests`, you can also specify a test-class or method:

```sh
source ./env/bin/activate
python -m twisted.trial tests.rest.admin.test_room tests.handlers.test_admin.ExfiltrateData.test_invite
```
```

Reviewer 2: and do `tox -e py36 -- test.rest.admin.test_room` etc. to the same regard

> The integration tests run the full Synapse, including your changes, to
> see if anything was broken. They are slower than the unit tests but will
> typically catch more errors.

Reviewer: ```suggestion
The integration tests run a full Synapse, including your changes, to
see if anything was broken. They are slower than the unit tests but will
typically catch more errors.
```

The following will run the test suite using postgres. You will need [Docker](https://docs.docker.com/get-docker/) installed.

```sh

> ./test_postgresql.sh

Reviewer: I don't think this is the integration tests? This looks to me like it is running the unit tests under PostgreSQL? I've never seen this file before though, good to know this is possible!

Author: I don't know, I've found this in the documentation, I haven't tested it myself!

license - in our case, this is almost always Apache Software License v2 (see
[LICENSE](LICENSE)).

# 2. What do I need?

The code of Synapse is written in Python 3. To do pretty much anything, you'll need [a recent version of Python 3](https://wiki.python.org/moin/BeginnersGuide/Download).

The source code of Synapse is hosted on Github. You will also need [a recent version of git](https://github.com/git-guides/install-git).

For some tests, you will need [a recent version of Docker](https://docs.docker.com/get-docker/).

## Under Unix (macOS, Linux, BSD, ...)

> Once you have installed Python 3, please open a Terminal and run:

Reviewer 3: I don't like to imply it *has* to be in `~/synapse/env`.

```suggestion
Once you have installed Python 3, open a Terminal and set up a virtualenv. For example:
```

```sh
mkdir -p ~/synapse
python3 -m venv ./env
source ./env/bin/activate
pip install -e ".[all,lint,mypy,test]"
pip install tox
```

This will install the developer dependencies for the project.

## Under Windows

TBD

# 3. Get the source.

The preferred and easiest way to contribute changes is to fork the relevant
project on github, and then [create a pull request](
https://help.github.com/articles/using-pull-requests/) to ask us to pull your
changes into our repo.

Please base your changes on the `develop` branch.

```sh
git clone git@github.com:YOUR_GITHUB_USER_NAME/synapse.git

> ```

Reviewer 3: bit odd that we tell people to base their changes on `develop`, and then tell them how to clone, but not how to switch branch.

Honestly I feel like telling people how to clone is out of scope here. If you need a "getting started with git" tutorial, you're going to need a lot more help than we're able to give in this document.

Author: This is meant as a subtle reminder that:
- they should clone with `git@` rather than `https`, because it just works better for most use cases;
- the should clone from their fork rather than `matrix-org/synapse`, because that will get messy when they try to open a PR.

directory. This documentation is intended primarily for sysadmins running their
own Synapse instance, as well as developers interacting externally with
Synapse. [docs/dev](docs/dev) exists primarily to house documentation for
Synapse developers. [docs/admin_api](docs/admin_api) houses documentation
regarding Synapse's Admin API, which is used mostly by sysadmins and external
service developers.

If you add new files added to either of these folders, please use [Github-Flavoured
Markdown](https://guides.github.com/features/mastering-markdown/).

Some documentation also exists in [Synapse's Github
Wiki](https://github.com/matrix-org/synapse/wiki), although this is primarily
contributed to by community authors.

# 7. Test, test, test!
<a name="test-test-test"></a>

While you're developing and before submitting a patch, you'll
want to test your code.

## Run the linters.

The linters look at your code and do two things:

- ensure that your code follows the coding style adopted by the project;
- catch a number of errors in your code.

They're pretty fast, don't hesitate!

```sh
source ./env/bin/activate
./scripts-dev/lint.sh
```

> Note that the linters *will modify your files* to fix styling errors.

Reviewer 3: No, the linters don't. The script does. (It's possible to run the linters outside the script).

```suggestion
Note that the this script *will modify your files* to fix styling errors.
```

Make sure that you have saved all your files.

If you wish to restrict the linters to only the files changed since the last commit
(much faster!), you can instead run:

```sh
source ./env/bin/activate
./scripts-dev/lint.sh -d
```

Or if you know exactly which files you wish to lint, you can instead run:

```sh
source ./env/bin/activate
./scripts-dev/lint.sh path/to/file1.py path/to/file2.py path/to/folder
```

## Run the unit tests.

The unit tests run parts of Synapse, including your changes, to see if anything
was broken. They are slower than the linters but will typically catch more errors.

```sh
source ./env/bin/activate

> python -m twisted.trial tests

Reviewer 3: you can also just write `trial tests`, fwiw...

```

If you wish to only run *some* unit tests, you may specify
another module instead of `tests` - or a test class or a method:

```sh
source ./env/bin/activate
python -m twisted.trial tests.rest.admin.test_room tests.handlers.test_admin.ExfiltrateData.test_invite
```

If your tests fail, you may wish to look at the logs:

```sh
less _trial_temp/test.log
```

## Run the integration tests.

The integration tests run a full Synapse, including your changes, to
see if anything was broken. They are slower than the unit tests but will
typically catch more errors.

> 

Reviewer 3: some misunderstanding here. `tox` runs the same set of tests as `trial`. I'm not entirely sure I'd describe those tests as integration tests.

Author: So, is there any way to run integration tests?

If not, I suppose that I should entirely remove this subsection.

Reviewer: Integration tests are run via sytest, we should probably point to the sytest repo that has it's own readme and not go into detail with it here.

Reviewer 3: The previous version of the doc did exactly that: https://github.com/matrix-org/synapse/pull/9281/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055L262-L265

Author: > Integration tests are run via sytest, we should probably point to the sytest repo that has it's own readme and not go into detail with it here.

I believe that it would be better if at least an example was in this document.

Now, I just need to understand that README.md :)

To run the entire test suite using Sqlite3, use the following command:

```sh
source ./env/bin/activate
tox -e py36
```

If you wish to only run *some* integration tests, you may
similarly specify a module, e.g.:

```sh
source ./env/bin/activate
tox -e py36 -- test.rest.admin.test_room
```

Or, to run the test suite using postgres, use the following command. You will need [Docker](https://docs.docker.com/get-docker/) installed.

```sh
source ./env/bin/activate
./test_postgresql.sh
```

# 8. Submit your patch.

Once you're happy with your patch, it's time to prepare a Pull Request.

To prepare a Pull Request, please:

1. verify that [all the tests pass](#test-test-test), including the coding style;
2. add a [changelog entry](#changelog);
3. [sign off](#sign-off) your contribution;
4. Push code to commits to your fork of Synapse:
```sh
git push origin develop

> ```

Reviewer 3: I would not necessarily recommend pushing to `develop`, even on your own forked repo. It makes it difficult for you to work on a separate set of changes while the first is still being reviewed.

Again, I'm not really sure that we should be trying to teach people how to make PRs here.

Author: Fair enough.

```

# 9. Submit your patch.

Once you're happy with your patch, it's time to prepare a Pull Request.

To prepare a Pull Request, please:

1. verify that [all the tests pass](#test-test-test), including the coding style;
2. [sign off](#sign-off) your contribution;
3. `git push` your commit to your fork of Synapse;
4. on Github, [create the Pull Request](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request);
5. add a [changelog entry](#changelog) and push it to your Pull Request;

> 6. on Github, request review from `matrix.org / Synapse Core`;

Author: Well, it's still information that it took me 2 months to find so I'll try and find a way to mention it :)

flag to `git commit`, which uses the name and email set in your
`user.name` and `user.email` git configs.

# 9. Turn feedback into better code.

Once the Pull Request is opened, you will see a few things:

1. our automated CI (Continuous Integration) pipeline will run (again) the linters, the unit tests, the integration tests and more;
2. one or more of the developers will take a look at your Pull Request and offer feedback.

From this point, you should:

1. Look at the results of the CI pipeline.
  - If there is any error, fix the error.
2. If a developer has requested changes, make these changes.
3. Create a new patch with the changes.
  - Please do NOT overwrite the history. New patches make the reviewer's life easier.

>   - Push this patch to your Pull Request.

Reviewer: Probably best to refer to these as commits, not patches.

It might be worth linking the bit about not overwriting history to docs/dev/git.md

4. Back to 1.

> If the CI pipeline detects errors, please fix them. If the developers suggest change, please fix them, too.

Reviewer: This line seems to repeat the above, not sure if that is on purpose.

flag to `git commit`, which uses the name and email set in your
`user.name` and `user.email` git configs.

# 10. Turn feedback into better code.

Once the Pull Request is opened, you will see a few things:

1. our automated CI (Continuous Integration) pipeline will run (again) the linters, the unit tests, the integration tests and more;
2. one or more of the developers will take a look at your Pull Request and offer feedback.

From this point, you should:

1. Look at the results of the CI pipeline.
  - If there is any error, fix the error.

> 2. If a developer has requested changes, make these changes.

Reviewer: ```suggestion
2. If a developer has requested changes, make these changes and let us know if it is ready for a developer to review again.
```

> 1. Look at the results of the CI pipeline.
>   - If there is any error, fix the error.
> 2. If a developer has requested changes, make these changes.
> 3. Create a new commit with the changes.
>   - Please do NOT overwrite the history. New commits make the reviewer's life easier.
>   - Push this commits to your Pull Request.
> 4. Back to 1.

Reviewer: The sub-bullets aren't rendering properly:

```suggestion
1. Look at the results of the CI pipeline.
   - If there is any error, fix the error.
2. If a developer has requested changes, make these changes.
3. Create a new commit with the changes.
   - Please do NOT overwrite the history. New commits make the reviewer's life easier.
   - Push this commits to your Pull Request.
4. Back to 1.
```

flag to `git commit`, which uses the name and email set in your
`user.name` and `user.email` git configs.

# 9. Turn feedback into better code.

Once the Pull Request is opened, you will see a few things:

1. our automated CI (Continuous Integration) pipeline will run (again) the linters, the unit tests, the integration tests and more;
2. one or more of the developers will take a look at your Pull Request and offer feedback.

From this point, you should:

1. Look at the results of the CI pipeline.
  - If there is any error, fix the error.
2. If a developer has requested changes, make these changes.
3. Create a new commit with the changes.

>   - Please [do NOT overwrite the history](docs/dev/git.md#on-keeping-the-commit-history-clean). New commits make the reviewer's life easier.

Reviewer 3: I'm not sure this document offers much advice to the contributor? I think it might be more confusing than useful.

Author: Agreed. Rolling back that change.
